### PR TITLE
sceNpTrophy: Fix sceNpTrophyRegisterContext for handler abortion

### LINKED ISSUE #9415: Regression: Implemented Trophy cause Splinter Cell Trilogy HD stuck at Intro
## Quick summary
sceNpTrophyAbortHandle and sceNpTrophyRegisterContext cause Playable status to Intro in Splinter Cell Trilogy HD

## Details
It loads intro (including movie) and automatically stuck at choose game screen. it shows "Please Wait".
works on rpcs3-v0.0.7-9298-71c4a2a1_win64
but not working one commit after that, rpcs3-v0.0.7-9299-7cc6881a_win64

_exact_ build (or commit) information that introduced the regression you're reporting.
rpcs3-v0.0.7-9299-7cc6881a_win64 
https://github.com/RPCS3/rpcs3/pull/7111 (Implement sceNpTrophyTerm resources deallocation, sceNpTrophyAbortHandle)
Screenshot 2020-12-11 183353

#### 6. Please provide your system configuration:
* OS Windows 10
* CPU Ryzen 5 3600
* GPU GTX Super 1660
* Driver version latest

#### Please include.
tried using default setting, just run it, it will stuck at opening screen, while one commit before able to play up the first Splinter Cell until finish
i think Elad tried to improve this commit on https://github.com/RPCS3/rpcs3/pull/7087 but still not solve the problem

### PR DESCRIPTION
**Pull request by elad335**
## Emulation Features
* Implement handler abortion during context registeration.

## Bugfixes
* Fixed recursive calls to sceNpTrophy functions by callback from registering context.
* Turns out the reader lock for context registeration was incorrect because ctxt->tropusr and trophy directory are being modified. made it an exclusive lock after callback.

Fixes #9415

### CODE REVIEW

**Comment by Nekotekina** [comment]
Using recursive mutex is considered a bad practice. Did you think of better workarounds?

**Comment by elad335** [comment]
A small hack can be made so it reruns all error checks and calls get_context_ex twice before and after first callback including the INITIALIZED check. (with unlock, callback(), relock mutex) Who knows maybe fw does sonething similar as well.
If you agree of course.

**Comment by Nekotekina** [comment]
Well, typical workflow with normal mutex is lock - update - unlock, where update doubtfully allows some arbitrary code such as a callback to execute. Something with relocking after callback seems more sound.

**Comment by elad335** [comment]
Done.

**Comment by elad335** [comment]
* Turns out the reader lock for this function was incorrect because ctxt->tropusr and trophy directory is being modified. made it an exclusive lock after callback.

**Comment by Nekotekina** [comment]
Wait a second, this is not sceNpTrophyAbortHandle right?

File: rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp [hunk 1]
```diff
@@ -529,13 +531,47 @@ error_code sceNpTrophyRegisterContext(ppu_thread& ppu, u32 context, u32 handle,
...
+
+	auto [ctxt2, error2] = trophy_manager->get_context_ex(context, handle);
+
+	if (error2)
+	{
+		// Recheck for any errors, such as if AbortHandle was called
+		return error2;
+	}
+
+	// Parandoid checks: context/handler could have been destroyed and replaced with new ones with the same IDs
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:563]:
```suggestion
	// Paranoid checks: context/handler could have been destroyed and replaced with new ones with the same IDs
```

Comment (elad335) [reply rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp]: Fixed.

**Comment by jokoputera** [comment]
this one fixed the Splinter Cell Trilogy stuck at intro

**Pull request merged**